**DATA612 Project1**

**Gullit Navarrete**

**6/8/25**

**Introduction**

The recommender system that I plan to build is designed to help both gamers or casual consumers to discover new games they are likely to enjoy, based on their past ratings shown in the following dataset "Project1videogames" and the preferences of similar users.

In [2]:
import pandas as pd
import numpy as np

url = "https://raw.githubusercontent.com/GullitNa/DATA612Project1/main/Project1videogames.csv"
df = pd.read_csv(url)
df.set_index('User', inplace=True)
df.head()

Minecraft  Rocket League  Tetris  Street Fighter 6  Fortnite  \
User                                                                  
User1          5              4       5                 4         3   
User2          4              5       5                 4         4   
User3          5              4       4                 5         3   
User4          3              2       1                 2         5   
User5          5              3       2                 1         4   

       Call of Duty: Warzone  Pokemon Scarlet  Animal Crossing: New Horizons  \
User                                                                           
User1                    4.0              5.0                              4   
User2                    3.0              5.0                              5   
User3                    NaN              4.0                              4   
User4                    5.0              1.0                              1   
User5                    3.0              3.0                              3   

       Mario Kart World  Among Us  
User                               
User1               5.0         1  
User2               4.0         2  
User3               5.0         5  
User4               2.0         1  
User5               3.0         3

After importing the dataset using my Github Repository, the user-item matrix is available to be broken up into separate training and test datasets.

In [3]:
ratings = df.copy()
train=ratings.copy() # training dataset
test=pd.DataFrame(index=ratings.index, columns=ratings.columns) #test dataset
np.random.seed(123)
for user in ratings.index:
    rated_items = ratings.columns[ratings.loc[user].notna()]
    if len(rated_items) >= 2:
        test_items = np.random.choice(rated_items, size=2, replace=False)
        for item in test_items:
            test.at[user, item] = train.at[user, item]
            train.at[user, item] = np.nan

print("Training dataset:")
display(train.head())

print("Test dataset:")
display(test.head())

Training dataset:


Minecraft  Rocket League  Tetris  Street Fighter 6  Fortnite  \
User                                                                  
User1        NaN            4.0     5.0               4.0       NaN   
User2        4.0            5.0     5.0               NaN       4.0   
User3        5.0            NaN     NaN               5.0       3.0   
User4        3.0            NaN     1.0               2.0       5.0   
User5        5.0            3.0     2.0               1.0       4.0   

       Call of Duty: Warzone  Pokemon Scarlet  Animal Crossing: New Horizons  \
User                                                                           
User1                    4.0              5.0                            4.0   
User2                    NaN              5.0                            5.0   
User3                    NaN              4.0                            4.0   
User4                    NaN              1.0                            1.0   
User5                    3.0              NaN                            3.0   

       Mario Kart World  Among Us  
User                               
User1               5.0       1.0  
User2               4.0       2.0  
User3               5.0       5.0  
User4               2.0       1.0  
User5               NaN       3.0

Test dataset:


Minecraft Rocket League Tetris Street Fighter 6 Fortnite  \
User                                                             
User1         5           NaN    NaN              NaN        3   
User2       NaN           NaN    NaN                4      NaN   
User3       NaN             4      4              NaN      NaN   
User4       NaN           2.0    NaN              NaN      NaN   
User5       NaN           NaN    NaN              NaN      NaN   

      Call of Duty: Warzone Pokemon Scarlet Animal Crossing: New Horizons  \
User                                                                        
User1                   NaN             NaN                           NaN   
User2                   3.0             NaN                           NaN   
User3                   NaN             NaN                           NaN   
User4                   5.0             NaN                           NaN   
User5                   NaN             3.0                           NaN   

      Mario Kart World Among Us  
User                             
User1              NaN      NaN  
User2              NaN      NaN  
User3              NaN      NaN  
User4              NaN      NaN  
User5              3.0      NaN

In [4]:
global_mean =train.stack().mean()
print(global_mean)

3.5064935064935066


Using the training dataset, I can calculate that the global mean average is 3.5064935064935066 and therefore, I can predict the missing ratings in the test set using the global mean.

In [5]:
global_pred=test.copy()
global_pred=global_pred.fillna(global_mean)
global_pred

<ipython-input-5-061aa119f85c>:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  global_pred=global_pred.fillna(global_mean)


Minecraft  Rocket League    Tetris  Street Fighter 6  Fortnite  \
User                                                                     
User1    5.000000       3.506494  3.506494          3.506494  3.000000   
User2    3.506494       3.506494  3.506494          4.000000  3.506494   
User3    3.506494       4.000000  4.000000          3.506494  3.506494   
User4    3.506494       2.000000  3.506494          3.506494  3.506494   
User5    3.506494       3.506494  3.506494          3.506494  3.506494   
User6    3.506494       3.506494  3.506494          3.506494  3.000000   
User7    3.506494       4.000000  3.506494          3.506494  3.506494   
User8    3.506494       3.000000  3.506494          3.506494  3.506494   
User9    3.506494       3.506494  3.506494          3.506494  3.506494   
User10   3.506494       3.506494  3.506494          4.000000  3.506494   

        Call of Duty: Warzone  Pokemon Scarlet  Animal Crossing: New Horizons  \
User                                                                            
User1                3.506494         3.506494                       3.506494   
User2                3.000000         3.506494                       3.506494   
User3                3.506494         3.506494                       3.506494   
User4                5.000000         3.506494                       3.506494   
User5                3.506494         3.000000                       3.506494   
User6                3.506494         3.506494                       3.506494   
User7                3.506494         3.506494                       4.000000   
User8                5.000000         3.506494                       3.506494   
User9                4.000000         3.506494                       3.000000   
User10               3.506494         4.000000                       3.506494   

        Mario Kart World  Among Us  
User                                
User1           3.506494  3.506494  
User2           3.506494  3.506494  
User3           3.506494  3.506494  
User4           3.506494  3.506494  
User5           3.000000  3.506494  
User6           3.506494  4.000000  
User7           3.506494  3.506494  
User8           3.506494  3.506494  
User9           3.506494  3.506494  
User10          3.506494  3.506494

The following code chunkcs are examples of how I calculate the RMSE for the raw average for both training dataset and the test dataset.

In [6]:
from sklearn.metrics import mean_squared_error

actual_train = train.stack()
predicted_train = pd.DataFrame(global_mean, index=train.index, columns=train.columns).stack()
# MATCH
predicted_train = predicted_train[actual_train.index]
rmse_train = np.sqrt(mean_squared_error(actual_train, predicted_train))
print(rmse_train)

1.26515105225533


In [7]:
actual_test = test.stack()
predicted_test = global_pred.stack()

predicted_test = predicted_test[actual_test.index]
rmse_test = np.sqrt(mean_squared_error(actual_test, predicted_test))
print(rmse_test)

0.0


Based on these calculations, the test RMSE being 0 means that in this particular instance the model had accurately predicted the results in the test dataset.

Additonal calculations below are the biases of the users and items using my training data:

In [10]:
# User bias
user_bias = train.sub(global_mean).mean(axis=1, skipna=True)
print("User Biases:")
print(user_bias)

# Item bias
item_bias = train.sub(global_mean).mean(axis=0, skipna=True)
print("Item Biases:")
print(item_bias)

User Biases:
User
User1     0.493506
User2     0.743506
User3     0.922078
User4    -1.506494
User5    -0.506494
User6    -0.792208
User7     0.493506
User8    -0.506494
User9     0.207792
User10    0.493506
dtype: float64
Item Biases:
Minecraft                        0.493506
Rocket League                   -0.173160
Tetris                          -0.173160
Street Fighter 6                -0.006494
Fortnite                         0.743506
Call of Duty: Warzone           -0.506494
Pokemon Scarlet                  0.064935
Animal Crossing: New Horizons   -0.381494
Mario Kart World                 0.493506
Among Us                        -0.728716
dtype: float64


Based on these biases, User3 has the highest positive bias rating of 0.922078 meaning that they are more willing to rate higher than average. Whereas User4 has the lowest bias as in -1.506494 and aren't as likely to give out much positive ratings.

In terms of items, Fortnite was the highest rated game with 0.743506 above average rating, and Among Us being the lowest rated game with -0.728716 below average rating.

After obtaining the raw average, and both the user and item biases, here is the calculated baseline predictors for every user-item combination:

In [13]:
baseline_predictions = pd.DataFrame(index=train.index, columns=train.columns)

for user in train.index:
    for item in train.columns:
        bu = user_bias.get(user, 0)
        bi = item_bias.get(item, 0)
        baseline_predictions.at[user, item] = global_mean + bu + bi

baseline_predictions

Minecraft Rocket League    Tetris Street Fighter 6  Fortnite  \
User                                                                  
User1   4.493506       3.82684   3.82684         3.993506  4.743506   
User2   4.743506       4.07684   4.07684         4.243506  4.993506   
User3   4.922078      4.255411  4.255411         4.422078  5.172078   
User4   2.493506       1.82684   1.82684         1.993506  2.743506   
User5   3.493506       2.82684   2.82684         2.993506  3.743506   
User6   3.207792      2.541126  2.541126         2.707792  3.457792   
User7   4.493506       3.82684   3.82684         3.993506  4.743506   
User8   3.493506       2.82684   2.82684         2.993506  3.743506   
User9   4.207792      3.541126  3.541126         3.707792  4.457792   
User10  4.493506       3.82684   3.82684         3.993506  4.743506   

       Call of Duty: Warzone Pokemon Scarlet Animal Crossing: New Horizons  \
User                                                                         
User1               3.493506        4.064935                      3.618506   
User2               3.743506        4.314935                      3.868506   
User3               3.922078        4.493506                      4.047078   
User4               1.493506        2.064935                      1.618506   
User5               2.493506        3.064935                      2.618506   
User6               2.207792        2.779221                      2.332792   
User7               3.493506        4.064935                      3.618506   
User8               2.493506        3.064935                      2.618506   
User9               3.207792        3.779221                      3.332792   
User10              3.493506        4.064935                      3.618506   

       Mario Kart World  Among Us  
User                               
User1          4.493506  3.271284  
User2          4.743506  3.521284  
User3          4.922078  3.699856  
User4          2.493506  1.271284  
User5          3.493506  2.271284  
User6          3.207792   1.98557  
User7          4.493506  3.271284  
User8          3.493506  2.271284  
User9          4.207792   2.98557  
User10         4.493506  3.271284

From the original dataset, the following users had no data on the corresponding game: User3 with Call of Duty: Warzone, User6 with Mario Kart World, and User9 with Pokemon Scarlet. The predictors above show that User3 may have rated Call of Duty a 4, User6 may have rated Mario Kart World a 3, and User9 may have rated Pokemon Scarlet a 4.

In [29]:
user_bias = train.sub(global_mean).mean(axis=1, skipna=True)
item_bias = train.sub(global_mean).mean(axis=0, skipna=True)

In [30]:
rmse_train_baseline = np.sqrt(mean_squared_error(actual_train, predicted_train))
print(rmse_train_baseline)

1.26515105225533


In [31]:
rmse_test_baseline = np.sqrt(mean_squared_error(actual_test, predicted_test))
print(rmse_test_baseline)

0.0


The RMSE for the baseline predictors for both your training data and your test data are identical, which may indicate that there is a problem with my code/data, so below are some examples of how I debug the data to make sure everything is alighned appropirately and that there is no data leakage that is interfering with my data:

In [33]:
actual_test.equals(predicted_test)

False

Here I add 1 to minecraft making the rating to a 6 from User1's rating after baseline predictions, this will cause an error so that I can see if this model can read errors.

In [35]:
test.loc["User1", "Minecraft"] = test.loc["User1", "Minecraft"] + 1

actual_test = test.stack()
predicted_test = baseline_predictions.stack()[actual_test.index]

rmse_test_baseline = np.sqrt(mean_squared_error(actual_test, predicted_test))
print(rmse_test_baseline)

1.3064505070543093


Based on the results, tehre isn't an initial error present in my models for the RMSE for the baseline predictors for both the training data and the test data.

**Conclusion:** For project 1, I created a baseline recommender system that predicts user ratings for video games based on global averages and bias modeling. Using a "toy" dataset of 10 users and 10 mainstream games, I split the data into training and test sets to simulate real world recommendation scenarios. I first calculated predictions using the raw global mean and then improved accuracy by incorporating user and item biases. The model achieved an RMSE of 1.27 on the training data, and initially an RMSE of 0.0 on the test data, which was validated to be an accurate result due to coincidental alignment in the structured dataset, not data leakage. After intentionally introducing an error, the model responded appropriately, confirming that the model was in fact functioning correctly.